test/extended/pods/pdb.go

```diff
@@ -0,0 +1,182 @@
+package pods
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"reflect"
+	"strings"
+
+	g "github.com/onsi/ginkgo"
+	oapi "github.com/openshift/api/config/v1"
+	"github.com/openshift/origin/pkg/test/ginkgo/result"
+	exutil "github.com/openshift/origin/test/extended/util"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+// getAllControlPlaneStaticPods gets the static pod
+func getAllControlPlaneStaticPods(kubeClient kubernetes.Interface) (sets.String, error) {
+	pod, err := kubeClient.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	var controlPlanePods = sets.NewString()
+	for _, pod := range pod.Items {
+		_, isStaticPod := pod.Annotations["kubernetes.io/config.mirror"]
+		_, isControlPlanePod := pod.Annotations["target.workload.openshift.io/management"]
+		if isStaticPod && isControlPlanePod && strings.Contains(pod.Namespace, "openshift-") {
+			staticPodName := strings.Split(pod.Name, "-ip")[0]
+			if staticPodName == "etcd" {
+				// etcd pod name needs to be appended with -quorum as it has quorum guard associated
+				staticPodName = staticPodName+"-quorum"
+			}
+			controlPlanePods.Insert(staticPodName)
+		}
+	}
+	return controlPlanePods, nil
+}
+
+var _ = g.Describe("[sig-arch] Managed cluster", func() {
+	defer g.GinkgoRecover()
+
+	oc := exutil.NewCLIWithoutNamespace("ensure-pdb")
+
+	g.It("all platform components should have PDBs configured", func() {
+		kubeClient := oc.AdminKubeClient()
+
+		deployments, err := kubeClient.AppsV1().Deployments("").List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			e2e.Failf("unable to list deployments: %v", err)
+		}
+		statefulsets, err := kubeClient.AppsV1().StatefulSets("").List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			e2e.Failf("unable to list statefulsets: %v", err)
+		}
+		// ovn-kube, openshift-controller-manager are running as DS.
+		daemonsets, err := kubeClient.AppsV1().DaemonSets("").List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			e2e.Failf("unable to list daemonsets: %v", err)
+		}
+
+		var items []runtime.Object
+		for i := range deployments.Items {
+			items = append(items, &deployments.Items[i])
+		}
+		for i := range statefulsets.Items {
+			items = append(items, &statefulsets.Items[i])
+		}
+		for i := range daemonsets.Items {
+			items = append(items, &daemonsets.Items[i])
+		}
+		// iterate over the references to find valid images
+		infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("unable to get infrastructure: %v", err)
+		}
+		var inHAMode bool
+		if infra.Status.InfrastructureTopology == oapi.HighlyAvailableTopologyMode {
+			inHAMode = true
```

Comment:
This seems like it should be orthogonal.  Can't we get at this just by ignoring `replicas < 2` workloads?  Because even if they allow surging, we don't surge on drains.